src/com/google/enterprise/adaptor/database/UniqueKey.java

```diff
@@ -160,6 +163,51 @@
     return aclSqlCols;
   }
 
+  /**
+   * Attempt to extract ColumnTypes from the supplied Map
+   * of column names to java.sql.Type
+   *
+   * @param typeMap Map of column names to Integer java.sql.Types
+   */
+  void addColumnTypes(Map<String, Integer> typeMap) {
+    for (Map.Entry<String, Integer> entry : typeMap.entrySet()) {
+      if (types.get(entry.getKey()) != ColumnType.UNKNOWN) {
+        continue;
+      }
+      ColumnType type;
+      switch (entry.getValue()) {
+        case Types.INTEGER:
+        case Types.SMALLINT:
+        case Types.TINYINT:
+          type = ColumnType.INT;
+          break;
+        case Types.BIGINT:
+          type = ColumnType.LONG;
+          break;
+        case Types.CHAR:
+        case Types.VARCHAR:
+        case Types.LONGVARCHAR:
+          type = ColumnType.STRING;
+          break;
+        case Types.DATE:
+          type = ColumnType.DATE;
+          break;
+        case Types.TIME:
+          type = ColumnType.TIME;
+          break;
+        case Types.TIMESTAMP:
+          type = ColumnType.TIMESTAMP;
+          break;
+        default:
+          String errmsg = "Invalid UniqueKey type '" + entry.getValue() + "'"
```

Comment:
I think s/type/SQL type/ and maybe replace the second sentence with something like: "Set an explicit key type in db.uniqueKey."